File: Ocean Rescue/chat1.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    const int INF = 1e9;

    while (T--) {
        int N, M;
        cin >> N >> M;

        vector<string> grid(N);
        queue<int> q;

        for (int r = 0; r < N; r++) {
            cin >> grid[r];

            for (int c = 0; c < M; c++) {
                if (grid[r][c] == 'B') {
                    q.push(r * M + c);
                }
            }
        }

        // Multi-source BFS:
        // all bases start at time 0.
        vector<int> dist(N * M, INF);

        // Mark every base as a starting point.
        queue<int> bfs;
        while (!q.empty()) {
            int id = q.front();
            q.pop();

            dist[id] = 0;
            bfs.push(id);
        }

        int dr[] = {-1, 1, 0, 0};
        int dc[] = {0, 0, -1, 1};

        int lastShip = 0;

        while (!bfs.empty()) {
            int id = bfs.front();
            bfs.pop();

            int r = id / M;
            int c = id % M;

            if (grid[r][c] == 'S') {
                lastShip = max(lastShip, dist[id]);
            }

            for (int k = 0; k < 4; k++) {
                int nr = r + dr[k];
                int nc = c + dc[k];

                if (nr < 0 || nr >= N || nc < 0 || nc >= M)
                    continue;

                if (grid[nr][nc] == '#')
                    continue;

                int nid = nr * M + nc;

                if (dist[nid] != INF)
                    continue;

                dist[nid] = dist[id] + 1;
                bfs.push(nid);
            }
        }

        cout << lastShip << '\n';
    }

    return 0;
}

File: Ocean Rescue/chat1.java
import java.io.*;
import java.util.*;

public class chat1 {

    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner();

        int T = fs.nextInt();
        StringBuilder out = new StringBuilder();

        int[] dr = {-1, 1, 0, 0};
        int[] dc = {0, 0, -1, 1};

        while (T-- > 0) {
            int N = fs.nextInt();
            int M = fs.nextInt();

            char[][] grid = new char[N][M];
            int[] dist = new int[N * M];

            Arrays.fill(dist, -1);

            ArrayDeque<Integer> q = new ArrayDeque<>();

            for (int r = 0; r < N; r++) {
                grid[r] = fs.next().toCharArray();

                for (int c = 0; c < M; c++) {
                    if (grid[r][c] == 'B') {
                        int id = r * M + c;
                        dist[id] = 0;
                        q.add(id);
                    }
                }
            }

            int answer = 0;

            // All bases are inserted initially.
            // Therefore this is a multi-source BFS.
            while (!q.isEmpty()) {
                int id = q.poll();

                int r = id / M;
                int c = id % M;

                if (grid[r][c] == 'S') {
                    answer = Math.max(answer, dist[id]);
                }

                for (int d = 0; d < 4; d++) {
                    int nr = r + dr[d];
                    int nc = c + dc[d];

                    if (nr < 0 || nr >= N || nc < 0 || nc >= M)
                        continue;

                    if (grid[nr][nc] == '#')
                        continue;

                    int nid = nr * M + nc;

                    if (dist[nid] != -1)
                        continue;

                    dist[nid] = dist[id] + 1;
                    q.add(nid);
                }
            }

            out.append(answer).append('\n');
        }

        System.out.print(out);
    }

    static class FastScanner {
        private final InputStream in = System.in;
        private final byte[] buffer = new byte[1 << 16];
        private int ptr = 0, len = 0;

        private int read() throws IOException {
            if (ptr >= len) {
                len = in.read(buffer);
                ptr = 0;

                if (len == -1)
                    return -1;
            }

            return buffer[ptr++];
        }

        String next() throws IOException {
            StringBuilder sb = new StringBuilder();
            int c;

            do {
                c = read();
            } while (c <= ' ');

            while (c > ' ') {
                sb.append((char) c);
                c = read();
            }

            return sb.toString();
        }

        int nextInt() throws IOException {
            return Integer.parseInt(next());
        }
    }
}